### PR TITLE
Create sparse Merkle tree and sub-component classes with interface & constructor - Closes #6487 #6488

### LINKED ISSUE #6487: Create sparse Merkle tree and sub-component classes with interface
### Description

Implement interface for classes `SparseMerkleTree` `Leaf` `Branch` `Empty` and their sub-functions.

### Motivation

* To implement sparse Merkle trees wee need these classes in `lisk-tree` library 

### Acceptance Criteria

* Class interfaces should be present
* Interfaces for `constructor` `getHash` `getValue` `getKey` `getLeft` `getRight` `update` and other functions wherever necessary should be present
* Add test skeletons

### Additional Information
Related issues : #6485  
We need to consider 2 approaches : Object vs class for nodes.

### PR DESCRIPTION
**Pull request by Incede**
### What was the problem?

This PR resolves #6487 #6488 

### How was it solved?

* Added interface for `SparseMerkleTree` and different types of nodes
* Implemented constructor and basic functions

### How was it tested?

Added test interface

### CODE REVIEW

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { EMPTY_HASH } from './constants';
+
+export class SparseMerkleTree {
+	private readonly _rootHash: Buffer;
+	private readonly _keyLength: number;
+	public constructor(rootHash?: Buffer, keyLength = 1) {
```
Comment (shuse2) [inline elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:20]: keyLength should be default to `36`? (moduleID (4) + prefix (2) + key (32))

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { EMPTY_HASH } from './constants';
+
+export class SparseMerkleTree {
+	private readonly _rootHash: Buffer;
```
Comment (shuse2) [inline elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:18]: this shouldt be `readonly` but maybe temporally?

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	private readonly _keyLength: number;
+	public constructor(rootHash?: Buffer, keyLength = 1) {
+		this._keyLength = keyLength;
+		if (rootHash) {
```
Comment (shuse2) [inline elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:22]:
```
  this._rootHash = rootHash ?? EMPTY_HASH;
```

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts (hunk 2 above)
Comment (Incede) [reply elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:18]: right, only temporarily.

File: elements/lisk-tree/src/sparse_merkle_tree/branch_node.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { Leaf } from './leaf_node';
+import { branchHash } from './utils';
+
+export class Branch {
+	private _left: Leaf;
+	private _right: Leaf;
+	private _hash: Buffer;
+	public constructor(left: Leaf, right: Leaf) {
+		this._left = left;
+		this._right = right;
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/branch_node.ts:24]: You can use parameter properties here if you prefer. 

https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties

File: elements/lisk-tree/src/sparse_merkle_tree/branch_node.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	public get hash() {
+		return this._hash;
+	}
+	public get left() {
+		return this._left;
+	}
+	public get right() {
+		return this._right;
+	}
+	public update(newChild: Leaf, side: boolean) {
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/branch_node.ts:37]: Use of `side` is not readable. 
```suggestion
	public update(newChild: Leaf, left: boolean) {
```

File: elements/lisk-tree/src/sparse_merkle_tree/empty_node.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { EMPTY_HASH } from './constants';
+
+export class Empty {
+	private readonly _hash: Buffer;
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/empty_node.ts:18]: You can directly assign`EMPTY_HASH` to the attribute directly.  if you prefer so. Then no need of constructor. 
```suggestion
	private readonly _hash: Buffer = EMPTY_HASH;
```

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+import { EMPTY_HASH } from './constants';
+
+export class SparseMerkleTree {
+	private readonly _rootHash: Buffer;
+	private readonly _keyLength: number;
+	public constructor(rootHash?: Buffer, keyLength = 36) {
+		this._keyLength = keyLength;
+		this._rootHash = rootHash ?? EMPTY_HASH;
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:22]: Can we really initialize tree with a given root hash? Is it not always computed value?

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	}
+	public get rootHash(): Buffer {
+		return this._rootHash;
+	}
+	public get keyLength(): number {
+		return this._keyLength;
+	}
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:29]: I found this pattern in above few classes as well. It seems you just want to have a `readonly public` property in such cases which only initialized in the constructor.

File: elements/lisk-tree/src/sparse_merkle_tree/utils.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { hash } from '@liskhq/lisk-cryptography';
+
+export const leafHash = (key: Buffer, value: Buffer): Buffer =>
+	hash(Buffer.concat([Buffer.from('00', 'hex'), key, value]));
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/utils.ts:18]: Please make `Buffer.from('00', 'hex')`  a constant `LEAF_HASH_PREFIX`

File: elements/lisk-tree/src/sparse_merkle_tree/utils.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+export const branchHash = (leftHash: Buffer, rightHash: Buffer): Buffer =>
+	hash(Buffer.concat([Buffer.from('01', 'hex'), leftHash, rightHash]));
```
Comment (nazarhussain) [inline elements/lisk-tree/src/sparse_merkle_tree/utils.ts:20]: Same here.

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts (hunk 7 above)
Comment (shuse2) [reply elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts:22]: i think in general, we always want to specify the root, and navigate from there.
ie) when block execution starts, `new SparseMerkleTree(lastBlockHeader.stateRoot)` and if it's empty, start from empty hash?

**Review by shuse2 — APPROVED** [review]
other than @nazarhussain 's comment, LGTM

**Review by nazarhussain — APPROVED** [review]

**Pull request merged**